Fix digest auth rest sensors

timmccor opened the pull request:
## Description:
Any REST sensor which is using DIGEST authentication will fail with the following exception stack:

```
2019-10-03 19:52:51 ERROR (MainThread) [homeassistant.components.sensor] rest: Error on device update!
Traceback (most recent call last):
  File "/usr/home/hass/env/lib/python3.7/site-packages/homeassistant/helpers/entity_platform.py", line 292, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/home/hass/env/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 441, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/home/hass/env/lib/python3.7/site-packages/homeassistant/components/rest/sensor.py", line 175, in update
    self.rest.update()
  File "/usr/home/hass/env/lib/python3.7/site-packages/homeassistant/components/rest/sensor.py", line 226, in update
    self._request, timeout=self._timeout, verify=self._verify_ssl
  File "/usr/home/hass/env/lib/python3.7/site-packages/requests/sessions.py", line 653, in send
    r = dispatch_hook('response', hooks, r, **kwargs)
  File "/usr/home/hass/env/lib/python3.7/site-packages/requests/hooks.py", line 31, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)
  File "/usr/home/hass/env/lib/python3.7/site-packages/requests/auth.py", line 247, in handle_401
    if self._thread_local.pos is not None:
AttributeError: '_thread._local' object has no attribute 'pos'
```

Fundamentally this is because the requests library stores a bunch of important information in thread local storage when the request is prepared, and this is needed when the thread is later used.

To fix this I've delayed preparing until we actually use the request - this does have the impact that we now prepare before every send rather than only once, but that doesn't appear to have much of a cost looking through the requests library and avoids this nasty issue.

**Related issue (if applicable):** fixes (closed aged issue) #13524


## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

timmccor committed:
Fix digest auth rest sensors

balloob commented on homeassistant/components/rest/sensor.py:
```diff
@@ -229,22 +229,22 @@ def __init__(
             with requests.Session() as sess:
                 response = sess.send(
-                    self._request, timeout=self._timeout, verify=self._verify_ssl
+                    self._request.prepare(), timeout=self._timeout, verify=self._verify_ssl
```
In this case `prepare()` won't save us anything. I suggest that we just call `getattr(sess, method)(url…)`

timmccor replied:
Sure. I was just trying to change as little code as possible - happy to refactor if you wish.

timmccor replied:
@balloob Would you like me to preserve existing behaviour for non-digest authentication? Presumably this will perform a tiny bit better even if it isn't significant.

timmccor committed:
Refactor to use request()

timmccor committed:
Fix black complaints

timmccor committed:
Don't rename data variable

Don't rename the data variable, appears several other sensors have been coded to rely on it

timmccor commented:
@balloob I'm going crazy looking at these tests...

They seem to be failing as a method of 'None' is passed to request(). This method came straight from the config, which should be setting a default of 'GET' - so I'm at a loss as to how this could have happened. Looking at the test output it seems that none of the default config which should be there is there for the binary sensor. But no failures from the 'normal' sensor...

RestData is a class within sensor, but binary_sensor refers to it. And both sensor and binary_sensor call it in exactly the same way with exactly the same config as far as I can see. So why does binary_sensor fail?!

I must have missed something obvious. Can you spot it?

timmccor committed:
Fix tests

test_setup_missing_schema:
Change the exception to check for to PlatformNotReady. With the change away from prepared statements, we no longer get the MissingSchema error during setup - we get it when we attempt to call the endpoint. Since the result is PlatformNotReady, which is what we want, and the error log clearly contains a note that the schema is incorrect I think this is fine.

test_setup_failed_connect & test_setup_timeout:
These aren't checking for minimum config, and they're not invoking the correct method to have the default config filled in. Therefore I've just added the correct minimum config so these can continue to test what they're there to test.

test_update_request_exception:
Testing a request exception with the assert on line 404! Excellent. Anyway, we've moved from using the requests Session object to the requests.request function - so the patch needed to be altered to ensure the RequestException was raised.

timmccor committed:
Remove no longer needed import

timmccor commented:
Failing tests are in another component - suggest this is ready to merge @balloob

homeassistant commented:
Hello @timmccor,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`05a013dca63ba211ffc992db93ae38e0b75a3df7`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;28153&#x2F;commits&#x2F;05a013dca63ba211ffc992db93ae38e0b75a3df7) - This commit has something that looks like an email address (timmccor@castle2.mccormickweb.co.uk). Maybe try linking that to GitHub?.
- [`a898555dbcf741129ddec71fad3eb63dabfccf6b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;28153&#x2F;commits&#x2F;a898555dbcf741129ddec71fad3eb63dabfccf6b) - This commit has something that looks like an email address (timmccor@castle2.mccormickweb.co.uk). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hello @timmccor,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`a898555dbcf741129ddec71fad3eb63dabfccf6b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;28153&#x2F;commits&#x2F;a898555dbcf741129ddec71fad3eb63dabfccf6b) - This commit has something that looks like an email address (timmccor@castle2.mccormickweb.co.uk). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

timmccor committed:
Fix binary sensor test in same way

balloob commented:
Bueno 🎉
